kafka-rest/src/test/java/io/confluent/kafkarest/integration/ClusterTestHarness.java

```diff
@@ -533,14 +527,36 @@ protected final void produceJsonMessages(List<ProducerRecord<Object, Object>> re
     props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList);
     props.setProperty(ProducerConfig.ACKS_CONFIG, "all");
     Producer<Object, Object> producer = new KafkaProducer<Object, Object>(props);
+
     for (ProducerRecord<Object, Object> rec : records) {
+      doProduce(producer, rec);
+    }
+    producer.close();
+  }
+
+  private <T, R extends Producer> void doProduce(R producer, ProducerRecord<T, T> rec) {
+    boolean sent = false;
+    int errorCount = 0;
+    while (!sent) {
       try {
         producer.send(rec).get();
+        sent = true;
       } catch (Exception e) {
-        fail("Couldn't produce input messages to Kafka: " + e);
+        errorCount++;
+        if (errorCount > RETRY_COUNT) {
+          fail("Couldn't produce input messages to Kafka: " + e);
```

Comment:
nit: Do we need to explicitly close the producer first?

Reply:
Hmm, good question. If it's a transient KafkaException  (eg dodgy network) we want to retry without a close (which is what I'd written to start with)

But tried coding don't close for a RetriableException and do close for the rest, and it just got a bit tangled, so I now close for all errors.